### PR TITLE
gateway: Unsubscribe from a topic

### PR DESCRIPTION
**Pull request by besedad**
Adds a handler for MQTT-SN's `UNSUBSCRIBE` message.

### CODE REVIEW

File: gateway/handler.go [hunk 1]
```diff
@@ -623,6 +633,46 @@ func (h *handler) handleSubscribe(ctx context.Context, snSubscribe *snMsgs.Subsc
...
+		var ok bool
+		topic, ok = h.predefinedTopics.GetTopicName(h.clientID, snUnsubscribe.TopicID)
+		if !ok {
+			return fmt.Errorf("Unknown topic id %d", snUnsubscribe.TopicID)
+		}
+		topicID = snUnsubscribe.TopicID
+	case snMsgs.TIT_SHORT:
+		topic = snMsgs.DecodeShortTopic(snUnsubscribe.TopicID)
+		// topicID remains zero.
+	}
```
Comment (mprymek) [inline gateway/handler.go:664]: I think all this logic is relevant to `SUBSCRIBE` only, it's not applicable to `UNSUBSCRIBE`. Let's discuss it personally...

File: gateway/unsubscribe_transaction.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// MQTT-SN SUBACK contains TopicID field. We must use this transaction to save
```
Comment (mprymek) [inline gateway/unsubscribe_transaction.go:1]: This is copy&paste error: `UNSUBACK` does not contain `TopicID`.

**Review by mprymek — CHANGES_REQUESTED** [review]
Beside the comments below, I think there's one fundamental misconception with this PR: 

GW must have `SUBSCRIBE`-related transaction because `SUBACK` contains `TopicID` which must be remembered between messages. Hence, there must be a transaction just for this purpose (it's a 1.2 spec bug, fixed in 2.0 spec).

This doesn't apply to `UNSUBSCRIBE`. It's 1. initialized by a client 2. there's no "state" to be remembered. Hence, I think the transaction is not needed at all.

Again, please let's discuss it personally in detail.

File: gateway/gateway_test.go [hunk 3]
```diff
@@ -883,6 +883,127 @@ func TestSubscribeQOS2Wildcard(t *testing.T) {
...
+	assert.Equal(topic, mqttSubscribe.Topics[0])
+
+	// GW <--UNSUBACK-- MQTT broker
+	mqttUnsuback := mqttPackets.NewControlPacket(mqttPackets.Unsuback).(*mqttPackets.UnsubackPacket)
+	mqttUnsuback.MessageID = mqttSubscribe.MessageID
+	stp.mqttSend(mqttUnsuback, false)
+
+	// client <--UNSUBACK-- GW
+	snUnsuback := stp.snRecv().(*snMsgs.UnsubackMessage)
+	assert.Equal(snUnsubscribe.MessageID(), snUnsuback.MessageID())
```
[the remark this answers is not in the document]
Comment (besedad) [reply gateway/gateway_test.go]: I've removed the publish and pubacks from the tests.

File: gateway/gateway_test.go [hunk 4]
```diff
@@ -883,6 +883,127 @@ func TestSubscribeQOS2Wildcard(t *testing.T) {
...
+	snRegack.SetMessageID(snRegister.MessageID())
+	stp.snSend(snRegack, false)
+
+	// client <--PUBLISH-- GW
+	snPublish := stp.snRecv().(*snMsgs.PublishMessage)
+	assert.Equal(topicID, snPublish.TopicID)
+	assert.Equal(snMsgs.TIT_REGISTERED, snPublish.TopicIDType)
+	assert.Equal(payload, snPublish.Data)
+	assert.Equal(qos, snPublish.QOS)
+	assert.Equal(false, snPublish.DUP())
```
[the remark this answers is not in the document]
Comment (besedad) [reply gateway/gateway_test.go]: Yeah, it's gone. See my previous comment.

File: gateway/unsubscribe_transaction.go (hunk 2 above)
Comment (besedad) [reply gateway/unsubscribe_transaction.go]: This comment is no longer relevant. The file in question has been deleted.

**Review by mprymek — APPROVED** [review]

**Pull request merged**